### PR TITLE
Rewrite how information about commands are handled

### PR DESCRIPTION
**Pull request by BradleyChatha**
TLDR;

* Add package jaster.cli.infogen which contains all the logic for gathering information about a command and its arguments
    * This allows an 'official' area to collect information about a command, without anything else built on top of it.
    * Action function generation has to be resolved here unfortunately, but it's a bit of a minor issue.
    * Class support is broken, but idk what to do about that when it's now due to a compiler/language limitation.
    * Majority of information is gathered without CTFE but now largely uses templates, which may or may not bloat file size and compile times, haven't really tested.
    * I **really** wanted CommandInfo to also keep along aliases to each parameter's symbol, and I was so so close until I got hit with a strange compiler error I couldn't fix. alas...
* Add `CommandParser` which is sort of a lite version of `CommandLineInterface`
    * It only handles parsing for a single command.
    * Uses the `Result` type for error management.
    * **Only** parses data into a command, does not understand anything about executing it.
    * Useful for when CommandLineInterface is too heavy-weight.
* Refactor CommandLineInterface to use the new CommandParser thing.
    * I'm not 100% sure if all the help text is looking as nice as before, but after 11 hours of working on this PR, I think I deserve a break >:3
* Add resultAssert, a basic assert helper that integrates with the Result type.
* Rename `@CommandRawArg` to `@CommandRawListArg` as the former sounds a bit confusing to me. Very pedantic breaking change though.
* Add the Pattern type because messing around with raw strings was a messy relic from JCLI's initial code.

In other words:

* infogen is used for gathering information about a single command.
* CommandParser is used for parsing data into a single command. Builds on top of infogen.
* CommandLineInterface provides parsing, dispatch, help text, and dependency injection for multiple commands. Builds on top of infogen, CommandParser, helptext.d, resolver.d, and jioc.

So responsibilities are sort of getting separated now, but CommandLineInterface still has room to be shrunk down a bit

TODO:

- [x] (Maybe) Move command help text generation into something separate from CommandLineInterface, as it can be useful for people not using it.

- [x] (Maybe) Modify CommandLineInterface to allow the user to provide an output sink, instead of being hardcoded to use stdout.

- [x] Document the infogen package

- [x] Document the CommandParser

- [x] Revise core.d to see what information is redundant, needs to be removed/added, etc.

- [x] Update README to reflect all the new changes, additions, and class breakage. Class breakage can probably be mitigated with mixin templates + structs.

- [x] Pray there's no major regressions.

If there's any more large overhauls needing to be done regarding how commands are handled, now's also the time to mention them.

At the very least, `core.d` isn't as much of a behemoth as it used to be.

Overall, this removes half of the power-share that CommandLineInterface had, so it's even easier to use the library if you just wanted some args parsed or if you want to also build some extra code based off of the information about a command.

The use case of the user building their own sort of `CommandLineInterface`, while keeping most of the UDAs and capabilities of ArgBinder in tact is also stronger.

If only I managed to get the alias forwarding to work, *then* there'd be a lot of possibilites, since user code could then also examine parameter UDAs easily...

### CODE REVIEW

File: source/jaster/cli/infogen/gen.d [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+    alias ActionFunc = actionFuncFromAction!(Action, CommandT, ArgT, ArgBinderInstance);
+    
+    // Get the arg group if one is assigned.
+    static if(hasUDA!(ArgT, CommandArgGroup))
+        enum Group = getSingleUDA!(ArgT, CommandArgGroup);
+    else
+        enum Group = CommandArgGroup.init;
+
+    // Determine existance and parse scheme traits.
+    enum Existance = determineExistance!(CommandT, typeof(ArgT), Action);
```
Comment (BradleyChatha) [inline source/jaster/cli/infogen/gen.d:126]: oh dear, it just hit me I've spelt Existence wrong.

**Comment by BradleyChatha** [comment]
Finally had the will to actually get this through. This will be merged on Wednesday if no major flaws/issues are found.
